src/utils/Math.vy

```diff
@@ -0,0 +1,132 @@
+# @version ^0.3.7
+"""
+@title Standard Mathematical Utility Functions
+@license GNU Affero General Public License v3.0
+@author pcaversaccio
+@notice These functions implement standard mathematical utility
+        functions that are missing in the Vyper language. If a
+        function is inspired by an existing implementation, it
+        is properly referenced in the function docstring.
+"""
+
+
+@external
+@pure
+def mul_div(x: uint256, y: uint256, denominator: uint256, roundup: bool) -> uint256:
+    """
+    @dev Calculates "(x * y) / denominator" in 512-bit precision,
+         following the selected rounding direction.
+    @notice The implementation is inspired by Remco Bloemen's
+            implementation under the MIT license here:
+            https://xn--2-umb.com/21/muldiv. Furthermore,
+            the rounding direction design pattern is inspired
+            by OpenZeppelin's implementation here:
+            https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/math/Math.sol.
+    @param x The 32-byte multiplicand.
+    @param y The 32-byte multiplier.
+    @param denominator The 32-byte divisor.
+    @param roundup The Boolean variable that specifies whether
+           to round up or not. The default `False` is round down.
+    @return uint256 The 32-byte calculation result.
+    """
+    # Handle division by zero.
+    assert denominator != empty(uint256), "Math: mul_div division by zero"
+
+    # 512-bit multiplication "[prod1 prod0] = x * y".
+    # Compute the product "mod 2**256" and "mod 2**256 - 1".
+    # Then use the Chinese Remainder theorem to reconstruct
+    # the 512-bit result. The result is stored in two 256-bit
+    # variables, where: "product = prod1 * 2**256 + prod0".
+    mm: uint256 = uint256_mulmod(x, y, ~empty(uint256))
```

Comment:
Current approach costs 421 gas per call.

If you use `max_value(uint256)` instead of `~empty(uint256)`, the cost per call will go down to 396 gas.

Reply:
Good catch - fixed via https://github.com/pcaversaccio/snekmate/commit/478d087829b9b71832a5ce4f03cd335f020e1333 (made you a co-author of the commit).